File: delete&earn.java
// Time Complexity : O(Max(N, Max(nums)))
// Space Complexity : O(Max(nums))
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
class Solution1 {
    public int deleteAndEarn(int[] nums) {
        
        //checking edge case
        if(nums == null || nums.length == 0)
            return 0;
        
        int max = Integer.MIN_VALUE;
        
        //Find max value in the nums array
        for(int num: nums){
            if(max < num)
                max = num;
        }
        
        //create a new array of max+1 length
        int[] arr = new int[max+1];
        
        //store the total of each value in its index
        for(int num: nums){
            arr[num] += num;
        }
        
        //set the first choose & don't choose value of first index
        int skip=0, choose=arr[0];
        
        //calucalte echoose & don't choose value for each index
        for(int i=1; i<arr.length; i++){
            int prevSkip = skip;
            skip = Math.max(skip, choose);
            choose = prevSkip + arr[i]; 
        }
        
        //return the max of choose & dont' choose of last index
        return Math.max(skip, choose);
    }
}

File: Delete&earn.py
#Time Complexity: O(n)+O(max(n))
#Space Complexity: O(n)
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        maxsize=0
        for i in range(len(nums)):
            maxsize=max(nums[i],maxsize)
        arr=[0]*(maxsize+1)
        for i in range(len(nums)):
            arr[nums[i]]=arr[nums[i]]+nums[i]
        prev=arr[0]
        curr=max(arr[0],arr[1])
        for i in range(2,len(arr)):
            temp=curr
            curr=max(temp,prev+arr[i])
            prev=temp
        return curr